tsdb/head.go

```diff
@@ -1298,34 +1556,42 @@ func (h *headChunkReader) Chunk(ref uint64) (chunkenc.Chunk, error) {
 	}
 
 	s.Lock()
-	c := s.chunk(int(cid))
+	c, garbageCollect := s.chunk(int(cid), h.head.chunkDiskMapper)
+	defer func() {
+		if garbageCollect {
+			c.chunk = nil
```

Comment:
hm , how does this work, if it is garbage collected here how would the caller read the data?

Reply:
I think the comments are confusing. Here we are garbage collecting the `c`, which is `*memChunk`, while the reader uses `c.chunk`. So setting c.chunk `nil` does not garbage collect `c.chunk` as it is referenced somewhere else already. Note that we are collecting `c` after `c.chunk` is already used somewhere.

Reply:
Does that make sense?

Reply:
hm , no sorry I didn't understand what you mean.

Reply:
Here `c` and `c.chunk` are 2 different things (basically 2 pointers). After calling `s.chunk()`, we are mainly using `c.chunk`, so we should not garbage collect `c.chunk`. But the struct that is holding it, `c`, can be garbage collected (added back to the pool). Garbage collecting it does not affect `c.chunk`, and setting `c.chunk = nil` also does not affect `c.chunk` as that means changing the pointer and not the value inside the struct pointed by the pointer.

Reply:
I get it now, this explains it very simple
https://play.golang.org/p/BX_zJ5_QaYx